src/librustc_parse/parser/ty.rs

```diff
@@ -276,8 +276,33 @@ impl<'a> Parser<'a> {
     }
 
     fn parse_borrowed_pointee(&mut self) -> PResult<'a, TyKind> {
-        let opt_lifetime = if self.check_lifetime() { Some(self.expect_lifetime()) } else { None };
+        let mut opt_lifetime =
+            if self.check_lifetime() { Some(self.expect_lifetime()) } else { None };
         let mutbl = self.parse_mutability();
+        if self.token.is_lifetime() && mutbl == Mutability::Mut && opt_lifetime.is_none() {
+            // A lifetime is invalid here: it would be part of a bare trait bound, which requires
+            // it to be followed by a plus, but we disallow plus in the pointee type.
+            // So we can handle this case as an error here, and suggest `'a mut`.
+            // If there *is* a plus next though, handling the error later provides better suggestions
+            // (like adding parentheses)
+            if !self.look_ahead(1, |t| t.is_like_plus()) {
+                let span = self.prev_token.span.to(self.token.span);
+                let lifetime = self.expect_lifetime();
+
+                let mut err = self.struct_span_err(span, "lifetime must precede `mut`");
+                if self.span_to_snippet(span).is_ok() {
```

Comment:
Is this check necessary?

Reply:
I assumed it might be necessary to prevent weird suggestions when macros were involved (per the [dev guide](https://rustc-dev-guide.rust-lang.org/diagnostics.html#error-messages)).

However even *with* this I was able to produce a bad suggestion in a macro:

```
macro_rules! mac {
    ($lt:lifetime) => {
        fn w<$lt>(x: &mut $lt i32) {}
    }
}

mac!('a);
```
gives

```
error: lifetime must precede `mut`
  --> <file>:3:23
   |
LL |         fn w<$lt>(x: &mut $lt i32) {}
   |                       ^^^^^^^ help: place the lifetime before `mut`: `'a mut`
...
LL | mac!('a);
   | --------- in this macro invocation
```

Note how it replaces `$lt` with an actual lifetime `'a`.

Anybody with experience in diagnostic suggestions know how this should be done?